reference-implementation/lib/readable-stream-abstract-ops.js

```diff
@@ -37,22 +37,87 @@ export function CallReadableStreamPull(stream) {
   return undefined;
 }
 
+export function ErrorReadableStream(stream, e) {
+  if (stream._readableStreamReader !== undefined) {
+    if (stream._state === 'waiting') {
+      stream._readableStreamReader._resolveReadyPromise(undefined);
+    }
+
+    // rs.ready() was pending because there was a reader.
+    stream._resolveReadyPromise(undefined);
+
+    stream._readableStreamReader._rejectClosedPromise(e);
+
+    stream._readableStreamReader._state = 'errored';
+
+    stream._readableStreamReader = undefined;
+  } else if (stream._state === 'waiting') {
+    stream._resolveReadyPromise(undefined);
+  }
+  stream._rejectClosedPromise(e);
+
+  stream._storedError = e;
+  stream._state = 'errored';
+
+  return undefined;
+}
+
 export function CloseReadableStream(stream) {
-  stream._state = 'closed';
+  if (stream._readableStreamReader !== undefined) {
+    if (stream._state === 'waiting') {
+      stream._readableStreamReader._resolveReadyPromise(undefined);
+    }
+
+    // rs.ready() was pending because there was a reader.
+    stream._resolveReadyPromise(undefined);
```

Comment:
I think there is (or if there isn't, should be) some duplicate code here with `releaseLock()`. Might be good to extract them out into a single abstract op?